src/util/string2int.cpp

```diff
@@ -89,3 +91,77 @@ unsigned long long int unsafe_string2unsignedlonglong(
 {
   return str2number<unsigned long long int>(str.c_str(), base, false);
 }
+
+template <typename do_conversiont>
+auto wrap_string_conversion(do_conversiont do_conversion)
+  -> optionalt<decltype(do_conversion())>
+{
+  try
+  {
+    return do_conversion();
+  }
+  catch(const std::invalid_argument &)
+  {
+    return nullopt;
+  }
+  catch(const std::out_of_range &)
+  {
+    return nullopt;
+  }
+}
+
+optionalt<int> string2optional_int(const std::string &str, int base)
+{
+  return wrap_string_conversion(
+    [&]() { return std::stoi(str, nullptr, base); });
+}
+
+optionalt<unsigned> string2optional_unsigned(const std::string &str, int base)
+{
+  return wrap_string_conversion([&]() {
+    if(str.find('-') != std::string::npos)
+    {
+      // stoul wraps around if given negative numbers,
+      // but since we need to narrow cast this leads to weird behavior
+      // so we disallow it here
+      throw std::out_of_range{"negative values not allowed"};
+    }
+    auto const ul_value = std::stoul(str, nullptr, base);
+    auto const u_value = narrow_cast<unsigned>(ul_value);
+    if(u_value == ul_value)
```

Comment:
This is not pretty. What's the difference with just returning the result of `numeric_cast<unsigned>` ?

Reply:
The difference is that without this check we get effectively "random" (i.e. implementation defined) results on overflow (consider 32 bit `unsigned`, and 64 bit `unsigned long` - now 2^32 i.e. `"4294967296"` is a perfectly valid `unsigned long`, casting to `unsigned` will turn it into `0`. The caller can no longer tell the difference between that and a 'valid' 0, so they'd effectively have to inspect the string themselves to make sure the function result isn't lying to them. The whole point of these functions is to avoid gotchas like that.

Reply:
Sorry I thought we had implemented a numeric_cast from integral type to integral type but apparently it's just from mp_integer (or exprt) to integral type.

Reply:
Ideally, it would be nice to extend numeric_cast and use it here, but it doesn't need to be in this PR.
You could write `if(ul_value <= std::numeric_limits<T>::max() && ul_value >= std::numeric_limits<T>::min())` before doing the conversion (I don't know if that's much better).